scipy/stats/_binomtest.py

```diff
@@ -313,16 +314,37 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
             # special case as shortcut, would also be handled by `else` below
             pval = 1.
         elif k < p * n:
-            i = np.arange(np.ceil(p * n), n+1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = n-binary_search_for_binom_tst(a_fn,d*rerr,np.ceil(p * n),n)+1
             pval = binom.cdf(k, n, p) + binom.sf(n - y, n, p)
         else:
-            i = np.arange(np.floor(p*n) + 1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = binary_search_for_binom_tst(a_fn,d*rerr,0,np.floor(p*n) + 1,True)+1
             pval = binom.cdf(y-1, n, p) + binom.sf(k-1, n, p)
 
         pval = min(1.0, pval)
 
     result = BinomTestResult(k=k, n=n, alternative=alternative,
                              proportion_estimate=k/n, pvalue=pval)
     return result
+
+
+def binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
```

Comment:
```suggestion
def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
```
Names of private functions should begin with an underscore. Please include at least a brief docstring to indicate what the inputs are and what the function does.

Can standard library functions be used, e.g. from `bisect`?

At what size does this start becoming faster than the linear search? If it's quite large, please switch between the original linear approach and this at a reasonable value.

Reply:
bisect seems to work on arrays (https://docs.python.org/3/library/bisect.html), which defeats the purpose because the whole point here is to get rid of the array and binary search the function. Will do some tests on speed, but I suspect this will never be slower. We've been using this at my work place for quite a while (that usage was what caused me to discover the array search in the original function first place) and its always blazingly fast. Will add example usage to the file itself. Can you tell me where the test cases for this function are?

Reply:
Ok. How about scipy.optimize.bisect (or other bracket method in scipy.optimize.root_scalar)? I'm just trying to avoid adding a new implementation of a bisection algorithm as there's probably something suitable already.

Reply:
Yeah, I was looking for an existing library too. Couldn't find one that fit all needs. 

This function should return not exact match, but ix such that binom.pmf(ix)<d<binom.pmf(ix+1)

For scipy.optimize.bisect, it is meant to work on real valued functions and returns floats (not indices as is required here). The float will need to be converted to the appropriate index. If we think about all the edge cases, etc. this will become more complex than the binary search itself.

Reply:
Tried scipy.optimize.bisect, didn't return expected result.
>>> n=10; p=0.5
>>> a_fn = lambda x1:binom.pmf(x1,n,p)
>>> _binary_search_for_binom_tst(a_fn,0.1,0,np.floor(n*p),True)
## returns 2
>>>b_fn = lambda x1: a_fn(x1)-0.1
optimize.bisect(b_fn,0,np.floor(n*p))
## returns 4.999999999998863

Reply:
The reason this doesn't work is that binom.pmf(2,n,p) is 0.04394531249999999 and binom.pmf(3,n,p) is 0.11718750000000014. But for everything in between; ex: binom.pmf(2.2,n,p) you get 0. This throws all those root finding and bisection methods off. You could try to "fix" this behavior, but the code will become uglier than the binary_search.